### ISSUE TITLE
Possible data poisoning

### ISSUE THREAD

**Issue by LetterN** [megascatterbomb/MegaAntiCheat#8]
I'm concerned about how MasterBase deals with possibly fake data, it would be possible to upload fake/modified/different demos on the backend to falsely mark other people as cheaters (even if it takes some effort). Authentication wouldn't work as you could pretend to be a MegaAntiCheat client uploading a "totaly real" demo

**Comment by FlamingLily** [comment]
I'm not wholly sure how TF2 Demos work but I'd assume that the program would need more than one demo to flag a cheater, and if one demo doesn't align with other demos from that same game (assuming demos can store game IDs or something along those lines), it would be discarded

**Comment by NathanX-S** [comment]
Data poisoning is a big concern of mine. It's possible to query the master server/individual servers, store that data somehow and verify that against demos maybe? But that only prevents certain acts.

**Comment by whosmatt** [comment]
I second this. 
**For every cheating player, there will be demos that indicate cheating and there will be demos that do not.** Either because of a lack of clear evidence or because the cheater simply toggled off for that match.  
So for a detected cheater, **some percentage of undetected demos have to be allowed**. 

Now if one of the many ill intentioned and technically capable cheat developers decide to poison the data with manipulated demo submissions, they could force a false positive on an innocent player.  
**As long as the demos are submitted by players rather than steams own servers, manipulated data can render this entire operation infeasible.**

**Comment by jacobbox** [comment]
So this by no means completely solves the problem but it may add further resistance to malicious actors.

Prerequisites
-----
- All users submitting evidence have a linked steam account.

Process
-----
- Certain accounts are implicitly trusted (think key devs, admins etc)
- Any account can register a level of trust they have for any other account (think high/mid/low)
- Accounts get a trust score via the PageRank algorithm (this is also nice as it's just matrix multiplication which is reasonably efficient)
- Evidence from accounts with higher trust scores are weighed more heavily

Slightly Rambling Discussion
-----
- So this doesn't prevent false evidence however over time trusted sources of evidence naturally form (the same way top results on google are usually what you want) so over time it will supress malicious actors.  
- When a bad actor is found they can be removed from the trust circle by seting all incoming trust values to 0 (thus supressing the value of it's evidence to 0 and any account it trusted stops getting boosted by it - this also means that it is immune to dragging accounts down by bad accounts marking innocent accounts as trusted).
- Some accounts can be artificially trusted by making every account trust it a little bit (same way the random surfer model in normal PageRank debilitates cycles).  This effectively means that these accounts act as a source of trust further reducing the effect of bad actors.
- As mentioned this doesn't solve everything - bad demos still need to be found by humans, however it means that over time bad actors have less power and when they are found any accounts that were trusted by a bad account can be easily untrusted again.

TLDR
-----

This is a proposed technical solution preserves the innocent/base/cheater model but allows everyone to submit evidence however it supresses the weighting of evidence from bad accounts

**Comment by NathanX-S** [comment]
> So this by no means completely solves the problem but it may add further resistance to malicious actors.
> 
> ## Prerequisites
> * All users submitting evidence have a linked steam account and signed (with public private key cryptography)
> 
> ## Process
> * Certain accounts are implicitly trusted (think key devs, admins etc)
> * Any account can register a level of trust they have for any other account (think high/mid/low)
> * Accounts get a trust score via the PageRank algorithm (this is also nice as it's just matrix multiplication which is reasonably efficient)
> * Evidence from accounts with higher trust scores are weighed more heavily
> 
> ## Slightly Rambling Discussion
> * So this doesn't prevent false evidence however over time trusted sources of evidence naturally form (the same way top results on google are usually what you want) so over time it will supress malicious actors.
> * When a bad actor is found they can be removed from the trust circle by setting all incoming trust values to 0 (thus supressing the value of it's evidence to 0 and any account it trusted stops getting boosted by it - this also means that it is immune to dragging accounts down by bad accounts marking innocent accounts as trusted).
> * Some accounts can be artificially trusted by making every account trust it a little bit (same way the random surfer model in normal PageRank debilitates cycles).  This effectively means that these accounts act as a source of trust further reducing the effect of bad actors.
> * As mentioned this doesn't solve everything - bad demos still need to be found by humans, however it means that over time bad actors have less power and when they are found any accounts that were trusted by a bad account can be easily untrusted again.
> 
> ## TLDR
> This is a proposed technical solution preserves the innocent/base/cheater model and allows everyone to submit evidence however it supresses the weighting of evidence from manually identified bad accounts and boosts the evidence from community supported trusted accounts

Querying servers/master list may also provide a means to at least ascertain whether a game occurred with X players on Y IP at Z map, etc.

**Comment by Lelebees** [comment]
Please be aware that any voting system reliant entirely on the goodwill of people can, and will be used maliciously. jacobbox's [idea](https://github.com/megascatterbomb/MegaAntiCheat/issues/8#issuecomment-1593873574) could very well be misused and we'd get a botting problem for the anticheat, boosting someones trustfactor (or destroying it).

I think instead of relying on community opinion (which can be artificially influenced) the trust score should rely on the reports made by that account. i.e. the trust score calculation could rely on the number of correct reports made, or could be a ratio of wrong accusation to correct accusation (Although this proposed system is not very robust either, there would need to be some caveats as to how that data is interpreted and collected, but this is just an example)

**Comment by jacobbox** [comment]
So under my proposal if a given account is frequently giving wrong accusations it's trust score can be easily suppressed and I suspect any real implementation in production would need to support this however putting that aside: the community method I propose is far more resistant botting that you indicate:
1. Anti-cheat bots (for the purposes of this post simply referred to as bots) only have as much power as real humans have given them.  There is no difference between 1000 accounts with no trust trusting something and 1 account with no trust trusting someone.
2. It should be nearly impossible for bots to harm the trust scores of legitimate users as there is no way to indicate distrust.
3. A theoretical attack vector could occur by boosting a account with bot accounts however this would be horrifically inefficient,  The trust would have to flow from somewhere and if bots are able to automatically generate fake accusations demos that look correct then we have bigger problems that this proposal doesn't seek to solve.
4. Even if a attack to boost a users trust score occurred once uncovered it is very easy to undo (unless 1000's of real people have been tricked) by removing the few bot accounts that tricked people cutting off the trust to the entire bot network.

Proof
=====

Considering that I am making claims about a technical solution to the problem I have modelled it in an attempt to prove the accuracy of these claims.

The results from testing are as follows:

```
> Scenario 1: isolated bots
> admin   real1   real2   real3   bot1    bot2    bot3    bot4
> 0.59008 0.22128 0.09958 0.08907 0.0     0.0     0.0     0.0
> 
> Scenario 2: tricked human
> admin   real1   real2   real3   bot1    bot2    bot3    bot4
> 0.5563  0.20861 0.09388 0.08397 0.03485 0.00747 0.00747 0.00747
> 
> Scenario 3: tricked human & bot 4 promotion
> admin   real1   real2   real3   bot1    bot2    bot3    bot4
> 0.55767 0.20913 0.09411 0.08417 0.03788 0.0     0.0     0.01705
> 
> Scenario 4: tricked human & bot 4 promotion fan out/in
> admin   real1   real2   real3   bot1    bot2    bot3    bot4
> 0.55483 0.20806 0.09363 0.08375 0.03769 0.00565 0.00565 0.01074
```

In this example experiment literally half of the accounts are bot accounts (bot1-bot4) and there are 3 real users and a source of trust "admin".

In the first scenario all of the bots trust each other however no trusted human was tricked into trusting a bot.  As can be seen the bot accounts have 0 weight - they are completely untrusted and thus there is NO benefit to adding more accounts they remain untrusted.

In scenario 2 The trusted human "real3" is tricked into trusting bot1, Note that though bot1 gets a bit of trust most of the bot network still has less than 0.01 trust.

scenario 3 & 4 are both different strategy to attempt to generate a trusted account (bot4) via the bot network.  Note that the bot network can never get bot4 to more than 0.01705 trust without tricking another different trusted real human.  In this way bots are a very inefficient method to try and trick the trust-scorer and attacks like social engineering are a much more likely risk vector (as they would be with any other solution including the old one too).

The reason why in all of this testing bots are so bad at tricking the system is that at every layer of indirection a significant damping factor is applied returning a portion of the 'incoming' trust into trust for the original admin account.  In this way the only way to become more trusted is trick more people and those people can only trust you 45% (by default) as much at they are trusted (because damping factor) and thus the only reasonable attack vector that could cause substantial trouble is a highly trusted account being tricked.  This is very much akin to tricking a moderator to give you permissions - not unheard of but reasonably difficult.

Below is the script used to test and invite others to verify that i having missed anything in the code:

```py
import sys
import numpy as np

ADMIN_INDEX = 0


class Confidence:
    HIGH = 5
    MID = 3
    LOW = 1


def generate_trustworthiness(user_confidence_matrix: np.matrix, admin_trust_factor: float = 0.55):
    USER_COUNT = user_confidence_matrix.shape[0]

    # Ensure empty columns (sink nodes) trust only the admins
    column_with_only_zeros = np.all(user_confidence_matrix == 0, axis=0)
    for cix, is_zero in enumerate(column_with_only_zeros):
        if is_zero:
            user_confidence_matrix[ADMIN_INDEX, cix] = 1

    # ensure all columns sum to 1 (thus simulating page ranks link counting)
    column_sums = np.sum(user_confidence_matrix, axis=0)
    normalized_user_confidence = user_confidence_matrix / column_sums

    # Apply admin trust to generate iteration matrix
    forced_admin_confidence = np.zeros((USER_COUNT, USER_COUNT))
    forced_admin_confidence[ADMIN_INDEX, :] = 1

    iteration_matrix: np.matrix = (normalized_user_confidence * (1-admin_trust_factor)) + (forced_admin_confidence * admin_trust_factor)

    # Initialise trustworthiness
    trustworthiness = np.full((USER_COUNT, 1), 1/USER_COUNT)

    # Iterate algorithm
    for i in range(0, 100):
        trustworthiness = iteration_matrix @ trustworthiness

    # yield result
    return trustworthiness


# =================================================
# =================== TESTING =====================
# =================================================

users = ("admin", "real1", "real2", "real3", "bot1", "bot2", "bot3", "bot4")

empty_confidence_matrix = np.zeros((len(users), len(users)))

honest_confidence_matrix = empty_confidence_matrix.copy()
honest_confidence_matrix[users.index("real1"), users.index("admin")] = Confidence.HIGH  # The user real1 is trusted by the user admin a lot
honest_confidence_matrix[users.index("real3"), users.index("admin")] = Confidence.LOW  # The user real3 is trusted by the user admin a little
honest_confidence_matrix[users.index("real2"), users.index("real1")] = Confidence.MID  # The user real2 is trusted by the user real1 somewhat
honest_confidence_matrix[users.index("real3"), users.index("real2")] = Confidence.MID  # The user real3 is trusted by the user real2 somewhat


def pretty_print_trust(trust_matrix: np.matrix):
    np.savetxt(sys.stdout, np.vstack((users, trust_matrix.transpose().round(5))), fmt='%s', delimiter='\t')


# ================================================
# Scenario 1: isolated group of intra-trusting bots
# ================================================
scenario_1_matrix = honest_confidence_matrix.copy()

# all bots trust all other bots but no real users are tricked
scenario_1_matrix[users.index("bot1"), users.index("bot2")] = Confidence.HIGH
scenario_1_matrix[users.index("bot1"), users.index("bot3")] = Confidence.HIGH
scenario_1_matrix[users.index("bot1"), users.index("bot4")] = Confidence.HIGH
scenario_1_matrix[users.index("bot2"), users.index("bot1")] = Confidence.HIGH
scenario_1_matrix[users.index("bot2"), users.index("bot3")] = Confidence.HIGH
scenario_1_matrix[users.index("bot2"), users.index("bot4")] = Confidence.HIGH
scenario_1_matrix[users.index("bot3"), users.index("bot2")] = Confidence.HIGH
scenario_1_matrix[users.index("bot3"), users.index("bot1")] = Confidence.HIGH
scenario_1_matrix[users.index("bot3"), users.index("bot4")] = Confidence.HIGH
scenario_1_matrix[users.index("bot4"), users.index("bot2")] = Confidence.HIGH
scenario_1_matrix[users.index("bot4"), users.index("bot3")] = Confidence.HIGH
scenario_1_matrix[users.index("bot4"), users.index("bot1")] = Confidence.HIGH

print("\nScenario 1: isolated bots")
pretty_print_trust(generate_trustworthiness(scenario_1_matrix))

# ================================================
# Scenario 2: A human is tricked
# ================================================

scenario_2_matrix = scenario_1_matrix.copy()

scenario_2_matrix[users.index("bot1"), users.index("real3")] = Confidence.HIGH

print("\nScenario 2: tricked human")
pretty_print_trust(generate_trustworthiness(scenario_2_matrix))


# ================================================
# Scenario 3: A human is tricked & bots promote bot4 only
# ================================================

scenario_3_matrix = honest_confidence_matrix.copy()

# tricked human
scenario_3_matrix[users.index("bot1"), users.index("real3")] = Confidence.HIGH

# bot strategy
scenario_3_matrix[users.index("bot4"), users.index("bot2")] = Confidence.HIGH
scenario_3_matrix[users.index("bot4"), users.index("bot3")] = Confidence.HIGH
scenario_3_matrix[users.index("bot4"), users.index("bot1")] = Confidence.HIGH

print("\nScenario 3: tricked human & bot 4 promotion")
pretty_print_trust(generate_trustworthiness(scenario_3_matrix))


# ================================================
# Scenario 4: A human is tricked & bots promote bot4 via fan out fan in
# ================================================

scenario_4_matrix = honest_confidence_matrix.copy()

# tricked human
scenario_4_matrix[users.index("bot1"), users.index("real3")] = Confidence.HIGH

# bot strategy
scenario_4_matrix[users.index("bot2"), users.index("bot1")] = Confidence.HIGH
scenario_4_matrix[users.index("bot3"), users.index("bot1")] = Confidence.HIGH
scenario_4_matrix[users.index("bot4"), users.index("bot2")] = Confidence.HIGH
scenario_4_matrix[users.index("bot4"), users.index("bot3")] = Confidence.HIGH
scenario_4_matrix[users.index("bot4"), users.index("bot1")] = Confidence.HIGH

print("\nScenario 4: tricked human & bot 4 promotion fan out/in")
pretty_print_trust(generate_trustworthiness(scenario_4_matrix))
```

**Comment by xxmyn** [comment]
> > I'm concerned about how MasterBase deals with possibly fake data, it would be possible to upload fake/modified/different demos on the backend to falsely mark other people as cheaters (even if it takes some effort). Authentication wouldn't work as you could pretend to be a MegaAntiCheat client uploading a "totaly real" demo
> > Also how would MasterBase deal with multiple, differing demo (ie 1 client has a different demo, 6 others have near-same ones)? Would it be a "voting system" on the marking verdict?
> 
> Curious to know if such software currently exists? How does one edit a demo?

use a cheat to change a players variables before the demo records them for that tick, you could make people "flick" on your demo, fake them doing damage, anything, since it's all done on your client i'm also curious as to what would stop someone from just having a demo with a cheater in it and spoofing the players ID prior to uploading it, at which point the anticheat would mark the spoof target.
